test.py

```diff
@@ -1019,9 +1022,13 @@ def __init__(self, test_no: int, shortname: str, casename: str, suite) -> None:
     async def run(self, options: argparse.Namespace) -> Test:
 
         self._prepare_pytest_params(options)
+        architecture = self.suite.options.tmpdir.split(os.sep)[-1] if self.suite.options.tmpdir.split(os.sep)[-1] in ["x86_64", "aarch64"] else None
```

Comment:
We should not reverse dependencies like that.
Architecture of the solution should be layered and dependencies should go in one direction.

test.py doesn't know whether its being run from jenkins or locally by a developer or from some other application or whatever. The implementation of test.py must not depend on such things.

Reply:
As for now, test.py does not know such things, but it should if we want to add some features. Since the architecture is a part of a temporary directory parameter that CI passes to the test.py we need to know to be able to create directories in the correct place. Even for the creation of the directory we do not need this information, but we need to construct the URL that will be attached to the failed report as a link to the directory with all logs. As far as I remember this is what quite a lot of developers wanted.
I'll modify this logic to remove the architecture parameter.

Reply:
This can be done in test.py without introducing knowledge about the CI environment / reversed dependency.

For example, to determine the architecture, instead of parsing tmpdir path -- which depends on the specific fact that CI environment runs tests under a path which contains architecture in its name, and is not true in general (e.g. local developer runs) -- we can introduce a parameter to test.py, e.g. `--arch`, which Jenkins pipeline would then pass to test.py.

See also https://en.wikipedia.org/wiki/Dependency_injection

Reply:
In fact, for the particular case of architecture, we don't need a parameter...
https://docs.python.org/3/library/platform.html

Reply:
> we need to construct the URL that will be attached to the failed report as a link to the directory with all logs

Determining the URL belongs inside Jenkins pipeline logic, I believe.

test.py could then receive the URL as part of its configuration.

Reply:
we are still depending on what directory will be passed from CI and how it will be copied to artifacts later.
you are right that I do not need the architecture. I've modified already the logic of building the URL.

Reply:
we are still depending on what directory will be passed from CI and how it will be copied to artifacts later.
you are right that I do not need the architecture. I've modified already the logic of building the URL.

> URL

This is what I'm doing. I'm getting the BUILD_URL from environment variables that will give us the URL up to the build number and then construct the rest of the path.

Reply:
> This is what I'm doing. I'm getting the BUILD_URL from environment variables that will give us the URL up to the build number and then construct the rest of the path.

What about this:
```
        return f"<a href={build_url}artifact/{dir_path_relative}</a>"
```

Isn't the "artifact" directory name specific to Jenkins?

I'm not sure I understand why this "logs dir link" needs to be calculated inside test.py. Can't it be passed from Jenkins pipeline as parameter?


Also, why are you using environment variable, when most of test.py configuration is passed through commandline parameters or yaml configuration? We should minimize the number of different ways to pass parameters.

Reply:
> > This is what I'm doing. I'm getting the BUILD_URL from environment variables that will give us the URL up to the build number and then construct the rest of the path.
> 
> What about this:
> 
> ```
>         return f"<a href={build_url}artifact/{dir_path_relative}</a>"
> ```
> 
> Isn't the "artifact" directory name specific to Jenkins?
> 
> I'm not sure I understand why this "logs dir link" needs to be calculated inside test.py. Can't it be passed from Jenkins pipeline as parameter?
> 
> Also, why are you using environment variable, when most of test.py configuration is passed through commandline parameters or yaml configuration? We should minimize the number of different ways to pass parameters.

you are right, the artifacts directory is Jenkins-specific. but this property anyway needed only on the Jenkins report.
it should be done inside test.py because test.py is aware of the fact which test failed and what it should contain in that directory. 
It can be passed in jenkins pipeline. But why do we need an additional argument when we can extract it directly from the environment variable? Especially since it's a common way to get a build URL in Jenkins